File: web/src/api/sandbox.ts
import axios from 'axios'

const api = axios.create({
  baseURL: import.meta.env.VITE_API_URL || '/api/v1',
  timeout: 30000,
  withCredentials: true,
})

export interface Sandbox {
  id: string
  image: string
  cpu: string
  memory: string
  ttl: number
  status: string
  template?: string
  templateVersion?: number
  desired_state?: string
  lifecycle_status?: string
  status_reason?: string
  pod_phase?: string
  pod_ip?: string
  last_seen_at?: string
  created_at: string
  expires_at: string
  updated_at?: string
  deleted_at?: string
  persistence?: SandboxPersistence
  deletion?: SandboxDeletion
  runtimeKind?: string
  runtimeName?: string
  accessToken?: string
  accessUrl?: string
}

export interface SandboxDeletion {
  phase?: string
  startedAt?: string
  lastAttemptAt?: string
  nextRetryAt?: string
  attempts?: number
  forceLevel?: number
  lastError?: string
}

export interface SandboxPersistence {
  enabled: boolean
  mode?: string
  size?: string
  storageClassName?: string
  reclaimPolicy?: string
  volumeClaimName?: string
}

export interface SandboxMetadataListParams {
  id?: string
  template?: string
  desired_state?: string
  lifecycle_status?: string
  deletion_phase?: string
  created_from?: string
  created_to?: string
  deleted_from?: string
  deleted_to?: string
  page?: number
  page_size?: number
}

export interface SandboxMetadataListResponse {
  items: Sandbox[]
  total: number
  page: number
  page_size: number
}

export interface PVCMapping {
  pvcName: string
  namespace: string
  storageClassName?: string
  requestedSize?: string
  phase?: string
  pvName?: string
  sandboxId?: string
  sandboxLifecycleStatus?: string
  reclaimPolicy?: string
  state: 'bound' | 'orphan_pvc' | 'deleting' | 'dangling_metadata'
  source: 'db+k8s' | 'db' | 'k8s' | string
}

export interface PVCMappingListParams {
  sandbox_id?: string
  storage_class?: string
  state?: 'bound' | 'orphan_pvc' | 'deleting' | 'dangling_metadata'
  page?: number
  page_size?: number
}

export interface PVCMappingListResponse {
  items: PVCMapping[]
  total: number
  page: number
  page_size: number
}

export interface SandboxStatusHistoryItem {
  id: number
  sandbox_id: string
  source: string
  from_status: string
  to_status: string
  reason: string
  payload_json: string
  created_at: string
}

export interface SandboxStatusHistoryResponse {
  items: SandboxStatusHistoryItem[]
}

export interface ReconcileRun {
  id: string
  trigger_type: string
  started_at: string
  finished_at?: string
  total_db: number
  total_k8s: number
  drift_count: number
  fixed_count: number
  status: string
  error?: string
}

export interface ReconcileItem {
  id: number
  run_id: string
  sandbox_id: string
  drift_type: string
  action: string
  detail: string
  created_at: string
}

export interface ReconcileRunListResponse {
  items: ReconcileRun[]
}

export interface ReconcileRunDetailResponse {
  run: ReconcileRun
  items: ReconcileItem[]
}

export interface CreateSandboxRequest {
  template: string // required
  templateVersion?: number
  overrides?: {
    cpu?: string
    memory?: string
    ttl?: number
    env?: Record<string, string>
    persistence?: {
      size?: string
    }
  }
}

export interface ExecRequest {
  command: string[]
  timeout?: number
}

export interface ExecResponse {
  exit_code: number
  stdout: string
  stderr: string
}

export interface LogsResponse {
  logs: string
  events: string[]
}

export interface ActionResponse {
  message: string
}

export const sandboxApi = {
  list: (params?: { include_terminating?: boolean }) =>
    api.get<{ items: Sandbox[] }>('/sandboxes', { params }),

  listMetadata: (params?: SandboxMetadataListParams) =>
    api.get<SandboxMetadataListResponse>('/sandboxes/metadata', { params }),

  listPVCMappings: (params?: PVCMappingListParams) =>
    api.get<PVCMappingListResponse>('/sandboxes/pvcs', { params }),

  get: (id: string) => api.get<Sandbox>(`/sandboxes/${id}`),

  getStatusHistory: (id: string, params?: { limit?: number; before_id?: number }) =>
    api.get<SandboxStatusHistoryResponse>(`/sandboxes/${id}/status-history`, { params }),

  create: (data: CreateSandboxRequest) => api.post<Sandbox>('/sandboxes', data),

  delete: (id: string) => api.delete<Sandbox>(`/sandboxes/${id}`),

  restart: (id: string) => api.post<ActionResponse>(`/sandboxes/${id}/restart`),

  stop: (id: string) => api.post<ActionResponse>(`/sandboxes/${id}/stop`),

  start: (id: string) => api.post<ActionResponse>(`/sandboxes/${id}/start`),

  exec: (id: string, data: ExecRequest) => api.post<ExecResponse>(`/sandboxes/${id}/exec`, data),

  getLogs: (id: string) => api.get<LogsResponse>(`/sandboxes/${id}/logs`),

  uploadFile: (id: string, path: string, file: File) => {
    const formData = new FormData()
    formData.append('file', file)
    formData.append('path', path)
    return api.post(`/sandboxes/${id}/files`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    })
  },

  downloadFile: (id: string, path: string) =>
    api.get(`/sandboxes/${id}/files`, {
      params: { path },
      responseType: 'blob',
    }),

  triggerReconcile: () => api.post<ReconcileRunDetailResponse>('/sandboxes/reconcile'),

  listReconcileRuns: (limit = 20) =>
    api.get<ReconcileRunListResponse>('/sandboxes/reconcile/runs', { params: { limit } }),

  getReconcileRun: (id: string) =>
    api.get<ReconcileRunDetailResponse>(`/sandboxes/reconcile/runs/${id}`),
}

export default api

File: backend/pkg/model/sandbox.go
package model

import "time"

type SandboxStatus string

const (
	SandboxStatusPending     SandboxStatus = "pending"
	SandboxStatusRunning     SandboxStatus = "running"
	SandboxStatusSucceeded   SandboxStatus = "succeeded"
	SandboxStatusFailed      SandboxStatus = "failed"
	SandboxStatusStopped     SandboxStatus = "stopped"
	SandboxStatusTerminating SandboxStatus = "terminating"
	SandboxStatusUnknown     SandboxStatus = "unknown"
)

const (
	PVCMappingStateBound        = "bound"
	PVCMappingStateOrphanPVC    = "orphan_pvc"
	PVCMappingStateDanglingMeta = "dangling_metadata"
	PVCMappingSourceDBAndK8s    = "db+k8s"
	PVCMappingSourceDB          = "db"
	PVCMappingSourceK8s         = "k8s"
)

type Sandbox struct {
	ID              string              `json:"id"`
	Image           string              `json:"image"`
	CPU             string              `json:"cpu"`
	Memory          string              `json:"memory"`
	TTL             int                 `json:"ttl"`
	Env             map[string]string   `json:"env,omitempty"`
	Status          SandboxStatus       `json:"status"`
	Template        string              `json:"template,omitempty"`
	TemplateVersion int                 `json:"templateVersion,omitempty"`
	DesiredState    string              `json:"desired_state,omitempty"`
	LifecycleStatus string              `json:"lifecycle_status,omitempty"`
	StatusReason    string              `json:"status_reason,omitempty"`
	PodPhase        string              `json:"pod_phase,omitempty"`
	PodIP           string              `json:"pod_ip,omitempty"`
	LastSeenAt      *time.Time          `json:"last_seen_at,omitempty"`
	CreatedAt       time.Time           `json:"created_at"`
	ExpiresAt       time.Time           `json:"expires_at"`
	UpdatedAt       time.Time           `json:"updated_at"`
	DeletedAt       *time.Time          `json:"deleted_at,omitempty"`
	Persistence     *SandboxPersistence `json:"persistence,omitempty"`
	RuntimeKind     string              `json:"runtimeKind,omitempty"`
	RuntimeName     string              `json:"runtimeName,omitempty"`

	// Network access fields
	AccessToken string `json:"accessToken,omitempty"` // Access token for inbound requests
	AccessURL   string `json:"accessUrl,omitempty"`   // Base URL for accessing the sandbox
}

// SandboxPersistence describes effective persistence settings on a sandbox.
type SandboxPersistence struct {
	Enabled          bool   `json:"enabled"`
	Mode             string `json:"mode,omitempty"`
	Size             string `json:"size,omitempty"`
	StorageClassName string `json:"storageClassName,omitempty"`
	ReclaimPolicy    string `json:"reclaimPolicy,omitempty"`
	VolumeClaimName  string `json:"volumeClaimName,omitempty"`
}

// CreateSandboxRequest represents a request to create a sandbox from a template.
// All sandboxes must be created from a template.
type CreateSandboxRequest struct {
	// Template is required - all sandboxes must be created from a template
	Template        string            `json:"template" binding:"required"`
	TemplateVersion int               `json:"templateVersion"`
	Overrides       *SandboxOverrides `json:"overrides"`
}

// SandboxOverrides allows overriding template configuration
type SandboxOverrides struct {
	CPU         string                       `json:"cpu,omitempty"`
	Memory      string                       `json:"memory,omitempty"`
	TTL         *int                         `json:"ttl,omitempty"`
	Env         map[string]string            `json:"env,omitempty"`
	Persistence *SandboxPersistenceOverrides `json:"persistence,omitempty"`
	// Note: Network configuration cannot be overridden, it must be set in the template spec
}

// SandboxPersistenceOverrides allows selected persistence fields to be overridden per sandbox.
type SandboxPersistenceOverrides struct {
	Size string `json:"size,omitempty"`
}

type ExecRequest struct {
	Command []string `json:"command" binding:"required"`
	Timeout int      `json:"timeout"`
}

type ExecResponse struct {
	ExitCode int    `json:"exit_code"`
	Stdout   string `json:"stdout"`
	Stderr   string `json:"stderr"`
}

type SandboxListResponse struct {
	Items []Sandbox `json:"items"`
}

type PVCMappingListOptions struct {
	SandboxID    string
	StorageClass string
	State        string
	Page         int
	PageSize     int
}

type PVCMapping struct {
	PVCName               string `json:"pvcName"`
	Namespace             string `json:"namespace"`
	StorageClassName      string `json:"storageClassName,omitempty"`
	RequestedSize         string `json:"requestedSize,omitempty"`
	Phase                 string `json:"phase,omitempty"`
	PVName                string `json:"pvName,omitempty"`
	SandboxID             string `json:"sandboxId,omitempty"`
	SandboxLifecycleState string `json:"sandboxLifecycleStatus,omitempty"`
	ReclaimPolicy         string `json:"reclaimPolicy,omitempty"`
	State                 string `json:"state"`
	Source                string `json:"source"`
}

type PVCMappingListResponse struct {
	Items    []PVCMapping `json:"items"`
	Total    int          `json:"total"`
	Page     int          `json:"page"`
	PageSize int          `json:"page_size"`
}

type LogsResponse struct {
	Logs   string   `json:"logs"`
	Events []string `json:"events"`
}

// WSMessage represents a WebSocket message for interactive exec
type WSMessage struct {
	Type     string `json:"type"`               // "input", "output", "resize", "exit", "error"
	Data     string `json:"data,omitempty"`     // terminal data (input/output)
	Cols     int    `json:"cols,omitempty"`     // terminal columns (resize)
	Rows     int    `json:"rows,omitempty"`     // terminal rows (resize)
	ExitCode int    `json:"exitCode,omitempty"` // process exit code (exit)
	Message  string `json:"message,omitempty"`  // error message (error)
}

// ExecInteractiveRequest defines parameters for interactive exec
type ExecInteractiveRequest struct {
	Command []string `json:"command"`
	TTY     bool     `json:"tty"`
	Cols    int      `json:"cols"`
	Rows    int      `json:"rows"`
}
